# Homework 5 - Logistic Regression

## Your Name Here

Student Name:

Student UT EID:

---

Partner Name:

Partner UT EID:

---

Date Created:

Date Last Modified:

---

Totoal Points 20. 

In [101]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [105]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
# converting yes/no to 1/0 to use in analysis later
heart_df['Target'] = heart_df['Target'].map({'Yes':1, 'No': 0})
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal  Target  
0        2.3      3  0.0       fixed       0  
1        1.5      2  3.0      normal       1  
2        2.6      2  2.0  reversable       1  
3        3.5      3  0.0      normal       0  
4        1.4      1  0.0      normal       0  
..       ...    ...  ...         ...     ...  
298      1.2      2  0.0  reversable       1  
299      3.4      2  2.0  reversable       1  
300      1.2      2  1.0  reversable       1  
301      0.0      2  1.0      normal       1  
302      0.0      1  NaN      normal       0  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (5 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data. 

In [123]:
from sklearn.model_selection import train_test_split

# split target and predicting variables into training and testing datasets
target_train, target_test, x_multi_train, x_multi_test = train_test_split(heart_df['Target'], heart_df[['Age', 'Sex', 'Chol']], test_size=0.2, random_state=0)



74     1
153    1
64     1
296    1
287    0
Name: Target, dtype: int64 
      Age  Sex  Chol
74    44    1   197
153   55    1   289
64    54    1   188
296   59    1   176
287   58    1   220


# Task 2 - (10 points)

Generate a logistic regression model using your training data. 

Print out the accuracy if your model. 

In [125]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
model = LogisticRegression()

# fit the data to a logistic regressin model
model.fit(x_multi_train, target_train)

# use the logistic regression model to predict the target with testing dataset
# then calculate the accuracy score for the prediction
target_prediction = model.predict(x_multi_test)
accuracy = accuracy_score(target_test, target_prediction)

print("Accuracy score of logistic regression model:", accuracy)



Accuracy score of logistic regression model: 0.7540983606557377


# Task 3 - (5 points)


Generate the classification report for your Logistic regresion model and interpret your results regarding precision, recall and f1-score. 


In [130]:
from sklearn.metrics import classification_report

# calculate and print the classification report
cr = classification_report(target_test, target_prediction)
print(cr)

# Interpertation of results:
# Precision: 81% of predicted negatives for heart diseases were correct but only 69% of positives predicted were correct
# Recall: 74% of people without heart disease were predicted correctly and 77% of those with heart disease were predicted correctly
# F1-score: the harmoic mean of precision and recal is 78% for those without heart disease and 73% for those with heard disease

              precision    recall  f1-score   support

           0       0.81      0.74      0.78        35
           1       0.69      0.77      0.73        26

    accuracy                           0.75        61
   macro avg       0.75      0.76      0.75        61
weighted avg       0.76      0.75      0.76        61



# Task 4 - Optional Task (2 extra points)


Which other feature of your data can you use to improve your prediction? Build other models and improve your prediction. 
